src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/RazorCodeBlockFoldingProvider.cs

```diff
@@ -32,12 +35,71 @@ public async Task<ImmutableArray<FoldingRange>> GetFoldingRangesAsync(DocumentCo
 
                 // Directives remove the "@" but for collapsing we want to keep it for users.
                 // Shows "@code" instead of "code".
-                CollapsedText = "@" + codeBlock.DirectiveDescriptor.Directive 
+                CollapsedText = "@" + codeBlock.DirectiveDescriptor.Directive
             };
 
             builder.Add(foldingRange);
         }
 
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        var razorFileSyntaxWalker = new RazorFileFoldingRangeSyntaxWalker(codeDocument.Source);
+        razorFileSyntaxWalker.Visit(codeDocument.GetSyntaxTree().Root);
+        builder.AddRange(razorFileSyntaxWalker.Ranges);
+
         return builder.ToImmutableArray();
     }
+
+    private class RazorFileFoldingRangeSyntaxWalker : SyntaxWalker
```

Comment:
May as well move this to its own file. The folder only has folding related code anyway